### ISSUE TITLE
Deadlock in HTTP/2

### ISSUE THREAD

**Issue by ManickaP** [dotnet/runtime#48220]
When an HTTP/2 request is being completed and, at the same time, SETTING frame with SETTINGS_INITIAL_WINDOW_SIZE is received, a deadlock might occur.
Call stacks:
```
[Waiting on lock owned by Thread 64302] (Unknown Source:0)
System.Private.CoreLib.dll!System.Threading.Monitor.Enter(object obj, ref bool lockTaken) Line 47 (/home/manicka/Repositories/runtime/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs:47)
System.Net.Http.dll!System.Net.Http.Http2Connection.RemoveStream(System.Net.Http.Http2Connection.Http2Stream http2Stream) Line 1923 (/home/manicka/Repositories/runtime/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:1923)
System.Net.Http.dll!System.Net.Http.Http2Connection.Http2Stream.Complete() Line 351 (/home/manicka/Repositories/runtime/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:351)
System.Net.Http.dll!System.Net.Http.Http2Connection.Http2Stream.SendRequestBodyAsync(System.Threading.CancellationToken cancellationToken) Line 247 (/home/manicka/Repositories/runtime/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:247)
...
```
```
[Waiting on lock owned by Thread 64317] (Unknown Source:0)
System.Private.CoreLib.dll!System.Threading.Monitor.Enter(object obj, ref bool lockTaken) Line 47 (/home/manicka/Repositories/runtime/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs:47)
System.Net.Http.dll!System.Net.Http.Http2Connection.Http2Stream.OnWindowUpdate(int amount) Line 429 (/home/manicka/Repositories/runtime/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:429)
System.Net.Http.dll!System.Net.Http.Http2Connection.ChangeInitialWindowSize(int newSize) Line 692 (/home/manicka/Repositories/runtime/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:692)
System.Net.Http.dll!System.Net.Http.Http2Connection.ProcessSettingsFrame(System.Net.Http.Http2Connection.FrameHeader frameHeader, bool initialFrame) Line 632 (/home/manicka/Repositories/runtime/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:632)
System.Net.Http.dll!System.Net.Http.Http2Connection.ProcessIncomingFramesAsync() Line 355 (/home/manicka/Repositories/runtime/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:355)
...
```

**Comment by ManickaP** [comment]
Fixed in #47769

**Comment by karelz** [comment]
Reopening to track backport to 5.0

**Comment by ManickaP** [comment]
@karelz can we close this since the 5.0 PR https://github.com/dotnet/runtime/pull/47921 has beem merged?

**Comment by karelz** [comment]
Fixed in servicing branch for 5.0.5 (PR #47921) and in main branch for 6.0 (PR #47769).
